server/live-loan-router.js

```diff
@@ -0,0 +1,125 @@
+const fetch = require('./util/fetch');
+
+const drawLoanCard = require('./util/live-loan/live-loan-draw');
+
+function getRecommendedLoans(type, loginId, offset = 0, limit = 1) {
+	return new Promise((resolve, reject) => {
+		const endpoint = config.app.graphqlUri;
+		let query;
+		if (type === 'url') {
+			query = `{
+				ml {
+					recommendationsByLoginId(
+						segment: all
+						  loginId: ${loginId}
+						  offset: ${offset}
+						  limit: ${limit}
+					) {
+						values {
+							id,
+						}
+					}
+				}
+			}`;
+		} else if (type === 'img') {
+			query = `{
+				ml {
+					recommendationsByLoginId(
+						segment: all
+						  loginId: ${loginId}
+						  offset: ${offset}
+						  limit: ${limit}
+					) {
+						values {
+							name
+							id
+							borrowerCount
+							geocode {
+								country {
+									name
+								}
+							}
+							use
+							loanAmount
+							status
+							loanFundraisingInfo {
+								fundedAmount
+							}
+							image {
+								retina: url(customSize: "w960h720")
+							}
+						}
+					}
+				}
+			}`;
+		}
+
+		fetch(endpoint, {
+			method: 'POST',
+			headers: { 'Content-Type': 'application/json' },
+			body: JSON.stringify({
+				query
+			}),
+		})
+			.then(result => result.json())
+			.then(result => {
+				const loanData = get(result, 'data.ml.recommendationsByLoginId.values');
+				resolve(loanData);
+			}).catch(err => {
+				console.error(err);
+				reject(err);
+			});
+	});
+}
+
+module.exports = function liveLoanRouter() {
+	const router = express.Router();
+	const userRouter = express.Router({ mergeParams: true });
+	const urlRouter = express.Router({ mergeParams: true });
+	const imgRouter = express.Router({ mergeParams: true });
+
+	router.get('/', (req, res) => { res.sendStatus(500); });
+
+	router.use('/u', userRouter);
+	userRouter.get('/', (req, res) => { res.sendStatus(500); });
+	userRouter.get('/:userId', (req, res) => { res.sendStatus(500); });
+	userRouter.use('/:userId/url', urlRouter);
+	userRouter.use('/:userId/img', imgRouter);
+
+	// URL Router
+	urlRouter.get('/', (req, res) => { res.sendStatus(500); });
+	urlRouter.get('/:offset', async (req, res) => {
+		const { userId, offset } = req.params;
+		try {
+			const loanData = await getRecommendedLoans('url', userId/* , offset */);
+			const offsetLoanId = loanData[0].id;
+			res.redirect(302, `/lend/${offsetLoanId}`);
+		} catch (err) {
+			console.error(err);
+			res.redirect(302, '/lend-by-category/');
+			// res.sendStatus(500); // TODO, do we want 500 error here instead?
+		}
+	});
+
+	// IMG Router
+	imgRouter.get('/', (req, res) => { res.sendStatus(500); });
+	imgRouter.get('/:offset', async (req, res) => {
+		const { userId, offset } = req.params;
+		try {
+			const loanData = await getRecommendedLoans('img', userId/* , offset */);
+			const loanCardImgBuffer = await drawLoanCard(loanData[0]); // TODO: offset?
+			// TODO: Add loanCard img to memcache?
```

Comment:
Add loanCard img to memcache here?

Reply:
We need to come up with a system to reduce the number of images that might be cached. As I mentioned before, rounding the currently funded about to the nearest 20th percent would be a good start. Then you could have cache keys for images like so... `live-loan-##loanid##-0`, `live-loan-##loanid##-20`, `live-loan-##loanid##-40` etc...

This would at least give us a baseline of measurement to ensure we have caching instances that can hold enough entries.

Reply:
Then your process will be more like...
`
const loanCardCacheKey = live-loan-##loanid##-0
let cachedLoanCardImage = retrieveFromCache(live-loan-##loanid##-0)
if (!cachedLoanCardImage) {
  cachedLoanCardImage = await drawLoanCard()
}
... res.send(cachedLoanCardImage);
`

Reply:
We could also just cache one image for a loan, and use a ttl rather than a cache key based on the loan progress. (The cached value could be json, allowing you to store and check the time it was cached)

Reply:
@emuvente I'd be fine with that if we can build in a "delay" of sorts so that we're not re-generating the image for small changes in the funded amount. my interest is reduce re-generating the image repeatedly (aka load on the express object) and to prevent excessive cache entries for each loan...

Reply:
The ttl would handle that delay, as the image would not be regenerated while the cached image is young enough. It would also handle the excessive entries for each loan as only 1 image would ever be cached for a loan.

Reply:
@emuvente @mcstover To reiterate, we'll cache the generated image buffer to memcache for say 10(?) minutes, and serve that to anyone else who would see the same loan. We'll also add a max-age header for also 10(?) minutes to the image response.